src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/AbstractFrameCheck.java

```diff
@@ -0,0 +1,738 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2020 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.CheckUtil;
+import com.puppycrawl.tools.checkstyle.utils.ScopeUtil;
+
+/**
+ * Abstract class for checks which require to work with and identify many frames.
+ */
+@FileStatefulCheck
+public abstract class AbstractFrameCheck extends AbstractCheck {
```

Comment:
Do we have ability to avoid inheritance of Checks? and use composition. Can we look at this class as FrameHelper ?

Reply:
Sure we can, In fact this is what I was going for at the beginning, but then @strkkk  suggested me to consider using inheritance as it is done for `AbstractNameCheck`
In my opinion, the inheritance approach is more relevant  and convenient because in our case, not only we have common methods but also common fields.

Reply:
> AbstractNameCheck

it is very old design. We try to avoid inheritance as much as possible, 20 year ago engineers thought that that inheritance was the only way, a lot is changes from that time. 
There are bunch of old not very benefitical inheritances, but we can not fix it sometime and sometime it is already works ... so we keep it as is.

@strkkk , @pbludov , FYI.